cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -472,78 +526,95 @@ struct MapLookupFunctor {
     return Status::OK();
   }
 
-  static Status ExecMapArray(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  static Status ExecMapArray(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
     const auto& options = OptionsWrapper<MapLookupOptions>::Get(ctx);
-    const auto& query_key = options.query_key;
-    const auto& occurrence = options.occurrence;
-    const MapArray map_array(batch[0].array());
+    const UnboxedKey query_key = UnboxScalar<KeyType>::Unbox(*options.query_key);
+
+    const ArraySpan& map = batch[0].array;
+    const int32_t* offsets = map.GetValues<int32_t>(1);
+
+    // We create a copy of the keys array because we will adjust the
+    // offset and length for the map probes below
+    ArraySpan map_keys = map.child_data[0].child_data[0];
+    const ArraySpan& map_items = map.child_data[0].child_data[1];
+
+    std::shared_ptr<DataType> item_type =
+        checked_cast<const MapType*>(map.type)->item_type();
 
     std::unique_ptr<ArrayBuilder> builder;
-    if (occurrence == MapLookupOptions::Occurrence::ALL) {
-      RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(),
-                                list(map_array.map_type()->item_type()), &builder));
+    if (options.occurrence == MapLookupOptions::Occurrence::ALL) {
+      RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), list(item_type), &builder));
       auto list_builder = checked_cast<ListBuilder*>(builder.get());
       auto value_builder = list_builder->value_builder();
 
-      for (int64_t map_array_idx = 0; map_array_idx < map_array.length();
-           ++map_array_idx) {
-        if (!map_array.IsValid(map_array_idx)) {
+      for (int64_t map_index = 0; map_index < map.length; ++map_index) {
+        if (!map.IsValid(map_index)) {
           RETURN_NOT_OK(list_builder->AppendNull());
           continue;
         }
 
-        auto map = map_array.value_slice(map_array_idx);
-        auto keys = checked_cast<const StructArray&>(*map).field(0);
-        auto items = checked_cast<const StructArray&>(*map).field(1);
```

Comment:
Note: avoiding all this value slicing and array boxing is a huge efficiency bonus